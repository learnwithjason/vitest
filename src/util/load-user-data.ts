interface User {
  username: string;
  name: string;
  projects: string[];
  coolness?: number;
  favoriteFood: string;
}

async function loadUser(username: string): Promise<User | undefined> {
  const users: User[] = [
    {
      username: 'antfu7',
      name: 'Anthony',
      projects: ['vitest', 'vite'],
      favoriteFood: 'sushi',
    },
    {
      username: 'jlengstorf',
      name: 'Jason',
      projects: ['burgers', 'cheese'],
      favoriteFood: 'pizza',
    },
  ];

  return users.find((user) => user.username === username);
}

export async function loadUserData(username: string) {
  const user = await loadUser(username);

  if (!user) {
    throw new Error('no user found');
  }

  user.coolness = username === 'antfu7' ? 100 : -1;

  return user;
}
