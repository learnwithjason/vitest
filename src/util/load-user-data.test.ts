import { describe, test, expect } from 'vitest';
import { loadUserData } from './load-user-data';

describe('loadUserDetails', () => {
  test('loads user data as expected', async () => {
    const user = await loadUserData('antfu7');

    expect(user).toMatchInlineSnapshot(`
      {
        "coolness": 100,
        "favoriteFood": "sushi",
        "name": "Anthony",
        "projects": [
          "vitest",
          "vite",
        ],
        "snacks": true,
        "username": "antfu7",
      }
    `);
  });

  test('sets coolness level appropriately', async () => {
    const jason = await loadUserData('jlengstorf');
    const anthony = await loadUserData('antfu7');

    expect(jason.coolness).toBe(-1);
    expect(anthony.coolness).toBe(100);
  });

  test('throws an error for nonexistent users', async () => {
    expect(async () => await loadUserData('fakeuser')).rejects.toThrowError(
      'no user found',
    );
  });
});
