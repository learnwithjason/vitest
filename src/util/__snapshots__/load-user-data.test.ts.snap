// Vitest Snapshot v1

exports[`loadUserDetails > loads user data as expected 1`] = `
{
  "coolness": 100,
  "favoriteFood": "sushi",
  "name": "Anthony",
  "projects": [
    "vitest",
    "vite",
  ],
  "username": "antfu7",
}
`;
